File: 1745-palindrome-partitioning-iv/1745-palindrome-partitioning-iv.cpp
class Solution {
    int dp[2000][2000];
public:
    bool checkPartitioning(string s) {
        int n = s.length();
        for(int i = 0; i < n; ++i) dp[i][i] = 1;
        for(int i = 0; i + 1 < n; ++i) dp[i][i + 1] = (s[i] == s[i + 1]);
        for(int len = 3; len <= n; ++len)
            for(int i = 0; i + len - 1 < n; ++i)
                dp[i][i + len - 1] = (s[i] == s[i + len - 1] && dp[i + 1][i + len - 2]);
        
        for(int i = 1; i < n; ++i)
            for(int j = i; j + 1 < n; ++j)
                if(dp[0][i - 1] && dp[i][j] && dp[j + 1][n - 1]) return true;
        return false;
    }
};

File: 1745-palindrome-partitioning-iv/1745-palindrome-partitioning-iv.java
class Solution {
    public boolean checkPartitioning(String s) {
        int n = s.length();
        char[] A = s.toCharArray();
        Boolean[][] dp = new Boolean [n][n];
        
        for(int i = n - 1; i >= 0; --i)
            for(int j = i; j < n; ++j)
                if(A[i] == A[j]) dp[i][j] = ((i + 1 <= j - 1) ? dp[i + 1][j - 1] : true);
                else dp[i][j] = false;
        
        
        for(int i = 1; i < n; ++i)
            for(int j = i; j < n - 1; ++j)
                if(dp[0][i - 1] && dp[i][j] && dp[j + 1][n - 1]) return true;
        
        return false;
    }
}